### PR TITLE
[Concurrency] Compute the source of actor isolation in `ActorIsolationRequest`.

### PR DESCRIPTION
**Pull request by hborla**
Every time I see a new project failure that says `expression is ‘async’ but not marked with ‘await’` or other actor isolation error, and the callee is not `async` or explicitly annotated with a global actor attribute, I go on a hunt for where the isolation may have been inferred from. To save time searching for the source of actor isolation inference, this change records the source of actor isolation in `ActorIsolationRequest`.

For example:

```swift
@MainActor
protocol P {}

struct S: P {
  func f() {}
}

func call(s: S) {
  s.f()
}
```

The above code now produces the following diagnostics:

```
 error: call to main actor-isolated instance method 'f()' in a synchronous nonisolated context
 3 | 
 4 | struct S: P {
 5 |   func f() {}
   |        |- note: calls to instance method 'f()' from outside of its actor context are implicitly asynchronous
   |        `- note: main actor isolation inferred from conformance to protocol 'P'
 6 | }
 7 | 
 8 | func call(s: S) {
   |      `- note: add '@MainActor' to make global function 'call(s:)' part of global actor 'MainActor'
 9 |   s.f()
   |     `- error: call to main actor-isolated instance method 'f()' in a synchronous nonisolated context
10 | }
11 | 
```

The note `main actor isolation inferred from conformance to protocol 'P'` is new with this change.

This information is currently only used as an attached note to one error produced by the actor isolation checker, but this should be used in many more places. The source of actor isolation inference could also potentially be surfaced by SourceKit requests in the future.

Resolves: rdar://119905790

### CODE REVIEW

**Comment by hborla** [comment]
@swift-ci please smoke test

**Comment by hborla** [comment]
@swift-ci please smoke test

File: test/Concurrency/actor_isolation.swift [hunk 1]
```diff
@@ -1659,3 +1659,33 @@ actor SafeMutatingCall {
     _ = try await self.iterator.next()
   }
 }
+
+
+@MainActor
+func testLocalFunctoinIsolation() {
```
Comment (ktoso) [inline test/Concurrency/actor_isolation.swift:1665]: minor: don't let this hold off merging but typo here
```suggestion
func testLocalFunctionIsolation() {
```

**Review by ktoso — APPROVED** [review]
LGTM, very nice :)

File: test/Concurrency/actor_isolation.swift (hunk 1 above)
Comment (hborla) [reply test/Concurrency/actor_isolation.swift:1665]: Whoops, thanks! I'm going to merge this; I've got another diagnostics change locally and I'll fix it along with that.

Comment (hborla) [reply test/Concurrency/actor_isolation.swift:1665]: Fixed in https://github.com/swiftlang/swift/pull/75687

**Pull request merged**